dill/_dill.py

```diff
@@ -1493,8 +1525,20 @@ def save_type(pickler, obj):
        #print ("%s\n%s" % (obj.__bases__, obj.__dict__))
         for name in _dict.get("__slots__", []):
             del _dict[name]
-        pickler.save_reduce(_create_type, (type(obj), obj.__name__,
+        if pickler_is_dill:
+            pickler._recursive_cells[id(obj)] = []
+        name = getattr(obj, "__qualname__", obj.__name__)
+        pickler.save_reduce(_create_type, (type(obj), name,
                                            obj.__bases__, _dict), obj=obj)
+        if pickler_is_dill:
+            recursive_cells = pickler._recursive_cells.pop(id(obj))
```

Comment:
just in case, probably should provide a default here

Reply:
Here, I was using `pickler._recursive_cells` both as a lookup table and a stack. If the object is not the top-most entry on the stack, then the dictionary must have been tampered outside of the `save_type` function. I think throwing an exception instead of having a default value would be more reasonable.

Reply:
Ok, fair enough